text/0000-adaptive-hashing.md

```diff
@@ -0,0 +1,193 @@
+threshold.
+
+Fortunately, Robin Hood hashing can be easily extended with adaptive hashing.
+
+# Detailed design
+## The algorithm for adaptive hashing
+
+A HashMap with adaptive hashing has two states. One state is called “fast mode” and the other is
+“safe mode”. The fast mode is the inital state for HashMaps with keys of a type that can be hashed
+in one shot. Otherwise, a HashMap with complex keys is always in safe mode.  We switch to the safe
+mode when the following conditions are met:
+
+- an inserted entry's displacement >= 128, or the number of entries displaced by an inserted
+  entry >= 512
+- the load of the map is smaller than 20%
+- the map is in the fast mode
+
+The second condition reduces the odds of switching to safe hashing. The chance that the first
+condition is satisfied is tiny, and the chance that both are satisfied at the same time is
+negligible. Moreover, we add a flag to the map. The flag delays displacement reduction until the
+next insertion to make code simpler. Otherwise, rebuilding the map would invalidate our entry.
+The pseudocode for a function that replaces `insert` is:
+
+```
+fn safeguarded_insert(map, key):
+  entry = insert(map, key)
+  if the entry's displacement >= 128 or the number of entries displaced by entry >= 512:
+    set the flag for reducing displacement
+  return entry
+```
+
+Before the next insertion operation, the state must be checked. Conveniently, the `reserve` method
+is always called before insertion and entry search, so we add the following code to `reserve`:
+
+```
+fn reserve(map, ...):
+  if the flag for reducing displacement is set and the map uses fast hashing:
+    if the load of the map is higher than 20%:
+      grow the map
+    else:
+      switch the map's hash state to safe hashing
+      rebuild the map
+    clear the flag for reducing displacement
+  // ...
+```
+
+Here’s a state diagram for HashMap with adaptive hashing. The dashed edge means the state change is
+very unlikely, and the dotted edge means the state change is enormously unlikely.
+
+<img width="800" src="https://cdn.rawgit.com/pczarn/code/d62cd067ca84ff049ef196aa1b7773d67b4189d4/rust/robinhood/adaptive.svg">
+
+## Choosing constants
+
+The thresholds of 128 and 512 are chosen to minimize the chance of exceeding them. In particular, we
+want that chance to be less than 10^-8 with a load of 90% and less than 10^-30 with a load of 20%.
+For displacement, the smallest k that fits our needs is 90, so we round that up to 128. For the
+number of forward-shifted buckets, we choose k=512. Keep in mind that the run length is a sum of the
+displacement and the number of forward-shifted buckets, so its threshold is 128+512=640. Even though
+the probability of having a run length of more than 640 buckets may be higher than the probability
+we want, it should be low enough.
+
+<img width="600" src="https://cdn.rawgit.com/pczarn/code/d62cd067ca84ff049ef196aa1b7773d67b4189d4/rust/robinhood/lookup_cost.png">
```

Comment:
These might want to get encoded as base64-encoded data URIs, so the RFC is self-contained.